lib/vtls/gskit.c

```diff
@@ -855,7 +762,7 @@ static CURLcode gskit_connect_step1(struct connectdata *conn, int sockindex)
 
   /* Establish a pipelining socket pair for SSL over SSL. */
   if(conn->proxy_ssl[sockindex].use) {
-    if(inetsocketpair(sockpair))
+    if(Curl_socketpair(0, 0, 0, sockpair))
```

Comment:
This will make gskit fail on OS400:
- OS supports socketpair (thus `HAVE_SOCKETPAIR` should be defined), but only for `AF_UNIX` sockets (see https://www.ibm.com/support/knowledgecenter/en/ssw_ibm_i_72/apis/socketp.htm),
- Since it is available in OS, our own implementation will not be used.
- socketpair will then create `AF_UNIX` sockets which are incompatible with gskit (see https://www.ibm.com/support/knowledgecenter/en/ssw_ibm_i_72/apis/gsk_attribute_set_numeric_value.htm).

Changing the domain in the call to AF_INET will just make it fail.
If we want to avoid duplicating code, the real solution i(apart from not defining HAVE_SOCKETPAIR) is to keep an unconditional Curl_inetsocketpair() function that can always be called, while Curl_socketpair() can be conditional.

Reply:
Why? Curl_socketpair() should be possible to always be used as it will be a mere macro for those systems that don't need our own replacement.

It rather seems that `HAVE_SOCKETPAIR` should only be set if it can handle `AF_UNIX`  **and** such a socketpair is good enough on that system. So in the case you describe, we should rather remove the define?

Reply:
> So in the case you describe, we should rather remove the define?

Yes. But when the OS provides it, it is an unnatural solution and prone to future incorrect change.

Reply:
It's already an "unnatural" situation when the provided `socketpair()` implementation isn't usable, so it needs a special case nonetheless.